### PR TITLE
Allow flexible parameter shapes and multi-dimensional tape return shape in custom JVP

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
The custom JVP rule in `gradients/jvp.py` assumes tape parameters to be scalars, and allows for a single axis of tape return types.

**Description of the Change:**
This PR updates `compute_jvp_single` to allow for flexible parameter shapes and multi-dimensional tape return shapes.
It also boils down some code bits to be a bit more concise.

**Benefits:**
More flexible JVPs. In particular, the option to differentiate with the interfaces while using non-scalar differentiable tape parameters, and when using multi-dimensional tape returns like e.g. broadcasted `probs`

**Possible Drawbacks:**
Code complexity.
Also, we should do a benchmark, although I think the previously supported cases should not be touched too much in their essence.

**Related GitHub Issues:**
TBD

### CODE REVIEW

File: pennylane/gradients/jvp.py [hunk 1]
```diff
@@ -227,61 +256,46 @@ def jvp(tape, tangent, gradient_fn, shots=None, gradient_kwargs=None):
...
-        jvps = []
-        if multi_m:
-            for i in range(len_shot_vec):
-                jvps.append(compute_jvp_multi(tangent, jac[i]))
-        else:
-            for i in range(len_shot_vec):
-                jvps.append(compute_jvp_single(tangent, jac[i]))
-
-        return tuple(jvps)
+        return tuple(_jvp_fn(tangent, jac[i]) for i in range(len_shot_vec))
```
Comment (AlbertMitjans) [inline pennylane/gradients/jvp.py:298]: Thanks for cleaning this up! Now it looks much better 🔥

File: pennylane/gradients/jvp.py [hunk 2]
```diff
@@ -89,46 +74,91 @@ def compute_jvp_single(tangent, jac):
...
-        # With dimension e.g. probs
-        else:
-            jac = qml.math.reshape(jac, (1, -1))
-        res = qml.math.tensordot(jac, tangent, [[0], [0]])
-    # Multiple params
+        return qml.math.zeros((1, 0))
+
+    if single_param:
+        tangent = qml.math.stack(tangent)
+        num_l = len(tangent.shape[1:])
```
Comment (AlbertMitjans) [inline pennylane/gradients/jvp.py:138]: What is `num_l`?

File: pennylane/gradients/jvp.py [hunk 3]
```diff
@@ -89,46 +74,91 @@ def compute_jvp_single(tangent, jac):
...
-
-        # With dimension e.g. probs
-        else:
-            jac = qml.math.reshape(jac, (1, -1))
-        res = qml.math.tensordot(jac, tangent, [[0], [0]])
-    # Multiple params
+        return qml.math.zeros((1, 0))
+
+    if single_param:
+        tangent = qml.math.stack(tangent)
```
Comment (AlbertMitjans) [inline pennylane/gradients/jvp.py:137]: Do we need to stack if we just have one parameter?

File: pennylane/gradients/jvp.py (hunk 2 above)
Comment (AlbertMitjans) [reply pennylane/gradients/jvp.py:138]: Does it correspond to `L_i` in the comment? If so it would be nice to use the same name.

File: pennylane/gradients/jvp.py [hunk 4]
```diff
@@ -89,46 +74,91 @@ def compute_jvp_single(tangent, jac):
...
+        num_l = len(tangent.shape[1:])
+        tangent = qml.math.reshape(tangent, (-1,))
+        prod_l = tangent.shape[0]
+        new_shape = (jac.shape if num_l == 0 else jac.shape[:-num_l]) + (prod_l,)
+        jac = qml.math.cast(qml.math.convert_like(jac, tangent), tangent.dtype)
+        jac = qml.math.reshape(jac, new_shape)
+        return qml.math.tensordot(jac, tangent, [[-1], [0]])
```
Comment (AlbertMitjans) [inline pennylane/gradients/jvp.py:144]: You can avoid a `reshape`:
```suggestion
        jac = qml.math.cast(qml.math.convert_like(jac, tangent), tangent.dtype)
        if num_l != 0:
            new_shape = (jac.shape[:-num_l]) + (prod_l,)
            jac = qml.math.reshape(jac, new_shape)
        return qml.math.tensordot(jac, tangent, [[-1], [0]])
```

File: pennylane/gradients/jvp.py [hunk 5]
```diff
@@ -89,46 +74,91 @@ def compute_jvp_single(tangent, jac):
...
+
+    if isinstance(tangent, (tuple, list)) and any(t.ndim > 0 for t in tangent):
+        num_l = [t.ndim for t in tangent]
+        tangent = [qml.math.reshape(t, (-1,)) for t in tangent]
+        prod_l = [t.shape[0] for t in tangent]
+        tangent = qml.math.hstack(tangent)
```
Comment (AlbertMitjans) [inline pennylane/gradients/jvp.py:150]: Will values in `tangent` ever have different shapes? Because we might get an error when stacking.

File: pennylane/gradients/jvp.py [hunk 6]
```diff
@@ -89,46 +74,91 @@ def compute_jvp_single(tangent, jac):
...
-        if jac[0].shape == ():
-            res = qml.math.tensordot(jac, tangent, 1)
-
-        # With dimension e.g. probs
-        else:
-            res = qml.math.tensordot(jac, tangent, [[0], [0]])
-    return res
+        num_l = [0] * len(tangent)
+        prod_l = [1] * len(tangent)
+        tangent = qml.math.stack(tangent)
```
Comment (AlbertMitjans) [inline pennylane/gradients/jvp.py:154]: Why this distinction? 

FYI: We can move the 1st and 3rd lines outside of the `if`/`else` statement

**Review by AlbertMitjans — COMMENTED** [review]
Good job! Added questions and suggestions.

File: pennylane/gradients/jvp.py [hunk 7]
```diff
@@ -89,46 +74,91 @@ def compute_jvp_single(tangent, jac):
...
     """
+    # There is a number of case distinctions in this code:
+    #  - The JVP may be for one (A) or multiple (B) parameters
+    #    Call the number of parameters k
+    #  - The number R of tape return type dimensions may be between 0 (i) and 3 (iv)
+    #    Call the return type dimensions r_j
+    #  - Each parameter may have an arbitrary number L_i>=0 of dimensions
+    #
+    # In the following, (a, b) denotes a tensor_like of shape (a, b) and [(a,), (b,)] /
+    # ((a,), (b,)) denotes a list / tuple of tensors with the indicated shapes, respectively.
```
Comment (AlbertMitjans) [inline pennylane/gradients/jvp.py:85]: Should we add this in the docstring?

Comment (dwierichs) [reply pennylane/gradients/jvp.py:85]: Probably.. Yeah :D

File: pennylane/gradients/jvp.py (hunk 3 above)
Comment (dwierichs) [reply pennylane/gradients/jvp.py:137]: Yes, in order to obtain a `tensor_like`. The `tangent` for example can come as a single-element list of a scalar, like `[np.array(0.5)]`, but we want to work with `tangent.shape` and `tangent.dtype` further down, and reshape it...

File: pennylane/gradients/jvp.py (hunk 2 above)
Comment (dwierichs) [reply pennylane/gradients/jvp.py:138]: It does, good call! So this would be called `L_1` here and just `L` further down. Is that okay as a variable name? :grimacing:

File: pennylane/gradients/jvp.py (hunk 4 above)
Comment (dwierichs) [reply pennylane/gradients/jvp.py]: Ah, yes! Nice :)

File: pennylane/gradients/jvp.py (hunk 5 above)
Comment (dwierichs) [reply pennylane/gradients/jvp.py:151]: Yes they do initially but two lines up we change them to all be 1D :)

Comment (dwierichs) [reply pennylane/gradients/jvp.py:151]: This makes sense because whatever the shapes, we always want to contract over _all_ dimensions of the tangent, so to us it might as well be a single 1d object.

File: pennylane/gradients/jvp.py (hunk 6 above)
Comment (dwierichs) [reply pennylane/gradients/jvp.py:155]: I think this distinction was left over from a way more complicated version of the code that I had inbetween.
So I think this might indeed be unnecessary. One question though: In the `else` clause we have `tangent` be a builtin sequence of scalars, which all would get reshaped to have shape `(1,)`. Do we want to do that extra work? I imagine this to be a very common use case of the function.

File: pennylane/gradients/jvp.py [hunk 8]
```diff
@@ -89,46 +74,92 @@ def compute_jvp_single(tangent, jac):
...
+
+    if single_param:
+        tangent = qml.math.stack(tangent)
+        num_l = len(tangent.shape[1:])
+        tangent = qml.math.reshape(tangent, (-1,))
+        prod_l = tangent.shape[0]
+        jac = qml.math.cast(qml.math.convert_like(jac, tangent), tangent.dtype)
+        if num_l != 0:
+            new_shape = (jac.shape[:-num_l]) + (prod_l,)
+            jac = qml.math.reshape(jac, new_shape)
```
Comment (dwierichs) [inline pennylane/gradients/jvp.py:144]: @AlbertMitjans this unfortunately actually didn't work: If `num_l` is `0`, we still need to add the `(prod_l,)` axis.
Basically this ternary operator just compensates for the fact that `[:0]` is not the same as `[:length-0]`.
Instead of the ternary operator I could write
```suggestion
        shape = jac.shape
        new_shape = shape[:len(shape)-num_l] + (prod_l,)
        jac = qml.math.cast(qml.math.convert_like(jac, tangent), tangent.dtype)
        jac = qml.math.reshape(jac, new_shape)
```
Is that better?

**Review by dwierichs — COMMENTED** [review]
Thanks for the suggestions @AlbertMitjans ! Working on implementing them.
Regarding the case distinction for the multi-param case, I think it is a bit of a matter of taste?

File: pennylane/gradients/jvp.py (hunk 6 above)
Comment (dwierichs) [reply pennylane/gradients/jvp.py:155]: I tried to make it work without the case distinction (note the `hstack` vs `stack`) but did not find a nice solution.
Happy about any suggestions you might have :)

File: pennylane/gradients/jvp.py (hunk 7 above)
Comment (dwierichs) [reply pennylane/gradients/jvp.py]: Done

**Comment by dwierichs** [comment]
Thanks @AlbertMitjans! I addressed all questions I think :) Some simplifications were possible but as far as I can see not all of the suggested ones. Happy to hear any other improvements that might be doable :)

File: doc/releases/changelog-dev.md [hunk 9]
```diff
@@ -425,9 +425,15 @@
 
 <h3>Improvements</h3>
 
+<<<<<<< HEAD
+* The custom JVP rules in PennyLane now also support non-scalar and mixed-shape tape parameters as
+  well as multi-dimensional tape return types, like broadcasted `qml.probs`, for example.
+  [(#3766)](https://github.com/PennyLaneAI/pennylane/pull/3766)
+=======
```
Comment (AlbertMitjans) [inline doc/releases/changelog-dev.md:432]: Careful with the merge conflicts!

File: pennylane/gradients/jvp.py (hunk 2 above)
Comment (AlbertMitjans) [reply pennylane/gradients/jvp.py]: Yep! This is good 😄

File: pennylane/gradients/jvp.py (hunk 5 above)
Comment (AlbertMitjans) [reply pennylane/gradients/jvp.py:187]: But will they always have the same size? If they don't, `tangent` will be a list of 1D arrays with different sizes...

File: pennylane/gradients/jvp.py (hunk 6 above)
Comment (AlbertMitjans) [reply pennylane/gradients/jvp.py]: Ah ok! No worries we can leave it like this 👍

File: pennylane/gradients/jvp.py (hunk 8 above)
Comment (AlbertMitjans) [reply pennylane/gradients/jvp.py]: Oops, sorry for this misunderstanding. Yes this is good 👍

File: doc/releases/changelog-dev.md [hunk 10]
```diff
@@ -429,9 +429,15 @@
...
```
Comment (AlbertMitjans) [inline doc/releases/changelog-dev.md:436]: Careful with the merge conflicts!

File: pennylane/gradients/jvp.py (hunk 5 above)
Comment (dwierichs) [reply pennylane/gradients/jvp.py:187]: Yep, they might, but that's not a problem, because we do `hstack`, not `stack` :)

File: doc/releases/changelog-dev.md (hunk 10 above)
Comment (dwierichs) [reply doc/releases/changelog-dev.md:436]: :grimacing: Thanks!

Comment (dwierichs) [reply doc/releases/changelog-dev.md]: Done.

**Review by AlbertMitjans — APPROVED** [review]
Thanks!! Lgtm 💯 🚀

**Review by rmoyard — COMMENTED** [review]
Great clean up and great addition :100: Some comments

File: pennylane/gradients/jvp.py [hunk 11]
```diff
@@ -88,47 +78,123 @@ def compute_jvp_single(tangent, jac):
...
+    .. details::
+        :title: Technical description
+        :href: technical-description
+
+        There are multiple case distinctions in this function, for particular examples see above.
+
+        - The JVP may be for one **(A)** or multiple **(B)** parameters. We call the number of
+          parameters ``k``
+
+        - The number ``R`` of tape return type dimensions may be between 0 and 3.
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/jvp.py:90]: probs + shot vector + broadcasting -> 3 :)

File: pennylane/gradients/jvp.py [hunk 12]
```diff
@@ -88,47 +78,123 @@ def compute_jvp_single(tangent, jac):
...
+             - ``((r_1,..,r_R),..,(r_1,..,r_R))``
+             - tensor return, ``k`` scalar parameters
+           * - ``[(l_1,..,l_{L_1}),..,(l_1,..,l_{L_k})]``
+             - ``((l_1,..,l_{L_1}),..,(l_1,..,l_{L_k}))``
+             - scalar return, ``k`` tensor parameters
+           * - ``[(l_1,..,l_{L_1}),..,(l_1,..,l_{L_k})]``
+             - ``((r_1,..,r_R, l_1,..,l_{L_1}),..,(r_1,..,r_R, l_1,..,l_{L_k}))``
+             - tensor return, ``k`` tensor parameters
+
+        Note that the return type dimensions always remain the same, whereas the
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/jvp.py:152]: Ah, I meant that they are the same between all entries of `jac`, whereas the `l_1...l_{L_j}` part of the shapes can differ. Will make that more clear, and make it a "footnote", so it is clear to which rows it applies :)

File: pennylane/gradients/jvp.py [hunk 13]
```diff
@@ -88,47 +78,123 @@ def compute_jvp_single(tangent, jac):
...
+        tangent = qml.math.reshape(tangent, (-1,))
+        prod_l = tangent.shape[0]
+        shape = jac.shape
+        new_shape = shape[: len(shape) - L_1] + (prod_l,)
+        jac = qml.math.cast(qml.math.convert_like(jac, tangent), tangent.dtype)
+        jac = qml.math.reshape(jac, new_shape)
+        return qml.math.tensordot(jac, tangent, [[-1], [0]])
+
+    L = [t.ndim for t in tangent]
+    if isinstance(tangent, (tuple, list)) and any(t.ndim > 0 for t in tangent):
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/jvp.py:184]: Ah, but sometimes it is a `tensor_like`, isn't it? :thinking:

File: pennylane/gradients/jvp.py [hunk 14]
```diff
@@ -88,47 +78,123 @@ def compute_jvp_single(tangent, jac):
...
+        tangent = [qml.math.reshape(t, (-1,)) for t in tangent]
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/jvp.py:185]: Ah, yes probably, thanks!

File: pennylane/gradients/jvp.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/jvp.py:152]: Changed it.

File: pennylane/gradients/jvp.py [hunk 15]
```diff
@@ -88,47 +78,123 @@ def compute_jvp_single(tangent, jac):
...
-        # With dimension e.g. probs
-        else:
-            jac = qml.math.reshape(jac, (1, -1))
-        res = qml.math.tensordot(jac, tangent, [[0], [0]])
-    # Multiple params
+        return qml.math.zeros((1, 0))
+
+    if single_param:
+        tangent = qml.math.stack(tangent)
+        L_1 = len(tangent.shape[1:])
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/jvp.py:174]: `first_tangent_ndim`

File: pennylane/gradients/jvp.py [hunk 16]
```diff
@@ -88,47 +78,123 @@ def compute_jvp_single(tangent, jac):
...
+        tangent = qml.math.reshape(tangent, (-1,))
+        prod_l = tangent.shape[0]
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/jvp.py:176]: `tangent_size`

File: pennylane/gradients/jvp.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/jvp.py:184]: Done.

[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/jvp.py:184]: I checked and this actually happens. The additional `isinstance` check allows us to avoid iterating over a one-dimensional tangent, e.g. a `torch.tensor` or so, and checking each entry to be non-scalar. However, if `tangent` is such a `tensor`, we already know it is not ragged and we can skip that iterative check.

File: pennylane/gradients/jvp.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/jvp.py:185]: Indeed, works :)

**Comment by rmoyard** [comment]
[sc-31372]

File: pennylane/gradients/jvp.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/gradients/jvp.py:186]: Ok thanks!

File: pennylane/gradients/jvp.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/gradients/jvp.py:90]: Thanks all good :+1:

**Review by rmoyard — APPROVED** [review]
Thanks @dwierichs great job :100:
